## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [1]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 42.3 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 146.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 290.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.9/16.9 MB 253.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.9/43.9 kB 278.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cuda-cupti-cu12==12.4.127; platform_system == "Linux" and platform_machine

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [2]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.23.2 pandas==1.5.3 tiktoken==0.6.0 pymupdf==1.25.1 langchain==0.1.17 langchain-core==0.1.51 langchain-community==0.0.36 chromadb==0.4.22 sentence-transformers==2.3.1 numpy==1.25.2 -q
!pip install posthog==3.0.1  # installing a compatible version

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 5.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.7/41.7 kB 3.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.7/41.7 kB 3.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 kB 3.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 kB 3.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 kB 3.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.2/40.2 kB 3.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.2/40.2 kB 3.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 3.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 4.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 3.2 MB

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [1]:
#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

In [2]:
# fuction for formatting response text
import textwrap
import re

def format_response_text(model_output, width=150):
    raw_text = model_output["choices"][0]["text"].strip()
    formatted_text = textwrap.fill(raw_text, width=width, replace_whitespace=False)
    verbose=False  # This hides prefix-match hit logs
    return formatted_text

## Question Answering using LLM

#### Downloading and Loading the model

In [3]:
# Defining the Hugging Face repository and model version for Mistral-7B fine-tuned
model_name_or_path = 'TheBloke/Mistral-7B-Instruct-v0.2-GGUF'

# Defining the file name for the quantized Mistral-7B model in GGUF format (Q6_K for optimal performance)
model_basename = 'mistral-7b-instruct-v0.2.Q6_K.gguf'

In [4]:
# Downloading the specified model file from Hugging Face Hub and store its local path
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename  # The specific model file to download (in GGUF format)
)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

In [82]:
# Loading the LLaMA model with specified context, GPU layers, and batch size
llm = Llama(
    model_path=model_path, #Path to the GGUF model file
    n_ctx=4096, #Sets the context window to 4096 tokens (how much text the model can "see" at once)
    n_gpu_layers=38, #Loads 38 model layers onto GPU for faster inference (set to 0 for CPU-only)
    n_batch=512, #Number of tokens processed at once
    seed=42  # Setting once here
)

AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [6]:
def response(query,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )
    return format_response_text(model_output)# calling function to format text

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [7]:
print(response('What is the protocol for managing sepsis in a critical care unit?'))

Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical
care unit. The following are general steps for managing sepsis in a critical care unit:

1. Early recognition: Recognize the signs and symptoms of
sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia,
tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.
2. Resusc


### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [8]:
print(response('What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'))

Llama.generate: prefix-match hit


Appendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the
abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:

1. Abdominal pain: The pain may start as a mild
discomfort around the navel or in the lower right abdomen, which then gradually moves to the right lower quadrant and becomes more severe over time.
The pain may be constant or intermittent and is often worsened by movement, coughing, or deep breathing


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [9]:
print(response('What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'))

Llama.generate: prefix-match hit


Sudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or
oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, or eyelashes.

The exact cause of
alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include
stress, genetics, viral infections, and certain medications.


### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [10]:
print(response('What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'))

Llama.generate: prefix-match hit


A person who has sustained a physical injury to brain tissue, also known as a traumatic brain injury (TBI), may require various treatments depending
on the severity and location of the injury. Here are some common treatments recommended for TBIs:

1. Emergency care: The first priority is to ensure
the person's airway is clear, they are breathing, and their heart is beating normally. In severe cases, emergency surgery may be required to remove
hematomas or other obstructions.
2. Medications: Depending on the symptoms, medications may be prescribed to manage conditions such as


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [11]:
print(response('What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'))

Llama.generate: prefix-match hit


First and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further
injury. Here are some necessary precautions:

1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and
prevent worsening the injury.
2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you
notice these symptoms, seek medical help immediately.
3. Immobilize the leg: Use a splint, sl


## Question Answering using LLM with Prompt Engineering

In [12]:
system_prompt = """
You are an AI assistant specializing in medical knowledge. Your role is to provide clear, precise, and medically reliable responses based on established medical guidelines and best practices.

When answering, prioritize factual correctness, align with widely accepted medical standards, and ensure clarity for both medical professionals and general users.
If a query requires specific reference materials beyond general medical knowledge, acknowledge the limitation rather than speculating.
"""

In [13]:
def generate_llama_response(query,max_tokens=128,temperature=0,top_p=0.95,top_k=50):

    # Create system message with instructions for the model
    instruction="Answer the following question clearly and concisely."
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=max_tokens,
        temperature=temperature,
        top_p=top_p,
        repeat_penalty=1.1,
        top_k=top_k,
        stop=['</s>'],
        echo=False,
    )

    return format_response_text(response)


### Query 1: What is the protocol for managing sepsis in a critical care unit?

**Response with basic parameters**

In [14]:
print(generate_llama_response('What is the protocol for managing sepsis in a critical care unit?'))

Llama.generate: prefix-match hit


The management of sepsis in a critical care unit involves early recognition, prompt initiation of antibiotics, fluid resuscitation, and supportive
measures. The Sequential [Sepsis-related] Organ Failure Assessment (SOFA) score can be used to identify patients at risk for developing sepsis or
worsening sepsis. Once sepsis is suspected, blood cultures should be obtained and broad-spectrum antibiotics administered based on the patient's
clinical presentation and local microbiology data. Fluid resuscitation with crystalloids or colloids may be necessary


**Response with Modified parameters - Temperature = 0.0**

In [17]:
def generate_llama_response1(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.0, # changed temparature
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=50,
        stop=['</s>'],
        echo=False,
        #seed=42,
    )

    # Return only the generated answer formatted text
    return format_response_text(response)

In [18]:
print(generate_llama_response1('What is the protocol for managing sepsis in a critical care unit?'))

Llama.generate: prefix-match hit


The management of sepsis in a critical care unit involves early recognition, prompt initiation of antibiotics, fluid resuscitation, and supportive
measures. The Sequential [Sepsis-related] Organ Failure Assessment (SOFA) score can be used to identify patients at risk for developing sepsis or
worsening sepsis. Once sepsis is suspected, blood cultures should be obtained and broad-spectrum antibiotics administered based on the patient's
clinical presentation and local microbiology data. Fluid resuscitation with crystalloids or colloids may be necessary to maintain adequate tissue
perfusion. Vasopressors and/or mechanical ventilation may also be required for hemodynamic support. Corticosteroids, anticoagulation, and other
adjunctive therapies may be considered based on individual patient needs and clinical trials data. Close monitoring of vital signs, laboratory values,
and organ function is essential to guide therapy and prevent progression to septic shock or multiple organ failure.


### **Observations on Q.1 responses:**

* **Response1** With Temp 0.7 is structured, Highly detailed and comprehensive,provides quantitative treatment goals for fluid resuscitation, oxygenation, and hemodynamic monitoring.Covers specific tools for diagnosis (SOFA, qSOFA, NEWS).
Includes specific targets (e.g., MAP ≥ 65 mmHg, ScvO₂ > 70%).
* **Response2** with Temp 0.0 is Written in paragraph form like a clinical summary or textbook explanation.Covers the same main concepts (early recognition, antibiotics, fluids, vasopressors, corticosteroids, supportive care).
Provides context for when to use different interventions (e.g., based on microbiology data or clinical trials).
* **Response1** contains more concrete clinical parameters and operational steps, making it more useful for procedural reference.
* **Response2** is better suited for educational summaries or high‑level discussions.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

 **Response with basic parameters**

In [19]:
print(generate_llama_response('What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'))

Llama.generate: prefix-match hit


Appendicitis is an inflammatory condition of the appendix, a small pouch located in the lower right side of the abdomen. Common symptoms include:
1.
Sudden pain in the lower right abdomen, often starting around the navel and moving to the right.
2. Loss of appetite.
3. Nausea and vomiting.
4. Fever
(often low-grade).
5. Abdominal swelling and tenderness.
6. Constipation or diarrhea.
7. Inability to pass gas or have a bow


**Recponse with modified parameter- Top_p = 0.85**

In [20]:
def generate_llama_response2(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.7,
        top_p=0.85,
        repeat_penalty=1.1,
        top_k=50,
        stop=['</s>'],
        echo=False,
        #seed=42,
    )

    # Return only the generated formatted answer text
    return format_response_text(response)


In [21]:
print(generate_llama_response2('What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'))

Llama.generate: prefix-match hit


Appendicitis is characterized by symptoms such as:
1. Sudden pain in the lower right abdomen, often starting around the belly button and then moving
to the right side.
2. Loss of appetite.
3. Nausea and vomiting.
4. Fever (usually low-grade).
5. Abdominal swelling and tenderness.

Appendicitis
cannot be cured via medicine alone, as the inflamed appendix may rupture if left untreated. If a rupture occurs, it can lead to peritonitis, a serious
infection in the abdominal cavity. Once the appendix has ruptured, emergency surgery called an appendectomy is necessary to remove the infected
appendix and prevent further complications. In some cases, a laparoscopic appendectomy may be performed, which uses small incisions and a camera to
guide the surgeon, resulting in a faster recovery time compared to traditional open surgery.


**Observation on Q.2 the responses**
* **Response1** generated with a top_p of 0.95, allows the model to sample from a broader range of potential next tokens. As a result, the response is more diverse and expansive, though its structure tends to be somewhat general.

* **Response2**, created with a lower top_p of 0.85, narrows the sampling range to more likely tokens. This leads to a response that is more focused, concise, and relevant.

Overall, **Response2** is preferable, as the reduced top_p value encourages the model to produce more targeted and coherent content, avoiding unnecessary variability.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

**Response with Basic Parameters**

In [22]:
print(generate_llama_response('What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'))

Llama.generate: prefix-match hit


Sudden patchy hair loss, also known as alopecia areata, is an autoimmune condition that results in hair loss from specific areas of the scalp or other
parts of the body. The exact cause of alopecia areata is unknown, but it's believed to be related to a problem with the immune system.

There are
several treatments for alopecia areata:

1. Corticosteroids: These medications can help reduce inflammation and suppress the immune response that
causes hair loss. They can be applied topically or taken orally.
2


**Response with modified parameter- max_token=512 and top_k=80**

In [23]:
def generate_llama_response3(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=512,
        temperature=0.7,
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=80,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return format_response_text(response)

In [24]:
print(generate_llama_response3('What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'))

Llama.generate: prefix-match hit


Sudden patchy hair loss can be distressing, but several potential causes and treatments exist. Here are some common causes and potential solutions:
1. Alopecia Areata: This is an autoimmune condition where your immune system attacks your hair follicles, causing circular bald patches. Treatment
options include topical corticosteroids, immunotherapy injections (such as Minoxidil or Diphencyprone), and in severe cases, systemic steroids or
other immune-modulating drugs.

2. Traction Alopecia: This occurs when hair is pulled too tightly and repeatedly, leading to damage and eventual hair
loss. Treatment involves stopping the damaging behavior and allowing the hair to grow back naturally.

3. Tinea Capitis (Ringworm of the scalp):
Fungal infection of the scalp that can cause circular patches of hair loss. Antifungal medication (topical or oral) is the treatment for this
condition.

4. Nutritional Deficiencies: Lack of adequate nutrients, such as iron, zinc, and vitamin B12, can lead to hair

**Observation on Q.3 responses**
* **Response1** keeps the cause list more focused and concise, ideal for general users,includes more medical treatment options (e.g., corticosteroid injections, LED therapy, immunomodulators).
* **Response1** is more structured for patient education or health blog content.
* **Response2** provides a wider range of potential causes, including autoimmune diseases, PCOS, hormonal imbalances, and genetics.
* **Response2** includes more medical treatment options (e.g., corticosteroid injections, LED therapy, immunomodulators), includes more medical treatment options (e.g., corticosteroid injections, LED therapy, immunomodulators).
* Combine both responses to form a well-rounded, layered explanation.
Use Response 1's structure, and enrich it with Response 2's medical detail.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

**Response with basic parameters**

In [25]:
print(generate_llama_response('What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'))

Llama.generate: prefix-match hit


The treatment for a brain injury depends on the severity and location of the injury. For mild injuries, such as concussions, rest, hydration, and
avoiding activities that may worsen symptoms are usually recommended. More severe injuries may require hospitalization, surgery, or rehabilitation.
Rehabilitation therapies, including physical therapy, occupational therapy, speech therapy, and cognitive rehabilitation, can help improve function
and reduce the impact of impairments on daily life. Medications may also be prescribed to manage symptoms such as pain, swelling, or seizures. In some
cases, assist


**Response with modified parameters-Top_k = 100**

In [26]:
def generate_llama_response4(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.7,
        top_p=0.95,
        repeat_penalty=1.1,
        top_k=100,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return format_response_text(response)


In [27]:
print(generate_llama_response4('What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'))

Llama.generate: prefix-match hit


The treatment for a traumatic brain injury (TBI) depends on the severity and location of the injury. Here are some common interventions:

1. Initial
care: The first priority is to ensure the person's airway is clear, they are breathing, and their circulation is stable. This may involve CPR or use
of a cervical collar to prevent neck movement.
2. Medications: Certain medications can help manage symptoms such as swelling in the brain (causing
increased pressure), seizures, and pain.
3. Surgery: Depending on the injury's location and severity, surgery may be necessary to remove hematomas or
decompress the skull.
4. Rehabilitation: Physical therapy, occupational therapy, speech therapy, and other forms of rehabilitation can help improve
function and mobility, as well as address cognitive deficits.
5. Assistive devices: Devices such as wheelchairs, walkers, or communication aids may be
required to help with daily activities.
6. Support groups: Joining support groups can provide emotional an

**Oberservation for Q.4 responses:**
* **Response1** (Top_k = 50) is more clinical, structured, and reads like a step-by-step recovery protocol — ideal for healthcare guidance.and Mentions lifestyle modifications (diet, stress management)
* **Response2** (Top_k = 100) is broader and more human-centered, offering a holistic overview with emotional and nutritional considerations — ideal for general education or patient-facing material and Mentions psychological counseling for both patients and families,Provides a strong conclusion emphasizing urgency and early care



### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [28]:
print(generate_llama_response('What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'))

Llama.generate: prefix-match hit


Firstly, if you suspect someone has fractured their leg while hiking, it's essential to keep them calm and still to prevent further injury. Try to
immobilize the affected leg using a makeshift splint or sling if possible. Do not attempt to move the person unless it is necessary for their safety
or to get them to medical help.

Once you have reached medical assistance, the healthcare professional will assess the severity of the fracture and
determine the appropriate treatment. This may include immobilizing the leg with a cast or splint, administering pain medication, or performing surgery
in severe


**Response with modified paramerters- Temp = 0.7, Top_p = 0.85, Top_k = 80,
Max_tokens = 256**

In [29]:
def generate_llama_response5(query, instruction="Answer the following question clearly and concisely."):

    # Create system message with instructions for the model
    system_message = f"[INST]<<SYS>>\n{instruction}\n<</SYS>>[/INST]"

    # Construct the final prompt using the user's query and system message
    prompt = f"{system_message} {query}"

    # Generate a response using the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=256,
        temperature=0.7,
        top_p=0.85,
        repeat_penalty=1.1,
        top_k=80,
        stop=['</s>'],
        echo=False,
    )

    # Return only the generated answer text
    return format_response_text(response)

In [30]:
print(generate_llama_response5('What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'))

Llama.generate: prefix-match hit


1. Assess the severity of the injury: If the fracture is open (compound), keep the area clean to prevent infection. Do not attempt to realign the bone
yourself if it is displaced or angulated severely.
2. Immobilize the leg: Use a splint, sling, or brace to immobilize the injured leg. Make sure the
injured limb is not bearing weight to avoid further damage.
3. Seek medical attention: It's essential to seek professional medical help as soon as
possible for proper diagnosis and treatment. If the injury occurs in a remote location, try to make your way to the nearest medical facility.
4. Pain
management: Over-the-counter pain relievers such as acetaminophen or ibuprofen can be used to manage pain. However, consult a healthcare professional
before taking any medication, especially if you're on other medications or have pre-existing conditions.
5. Proper positioning: Position the injured
leg in a comfortable and supported position during transportation. Keep it elevated above heart level to

**Observation on Q.5 responses-**
* **Response1** is clear and easy to follow but with lacks detailed emergency handling steps and It has missing preventive care aspects like clot prevention, muscle atrophy.
* **REsponse2** is More medically comprehensive, with Open vs. closed fracture handling, Use of splints, cleaning wounds, and not realigning bones, and Physical therapy and blood clot prevention.
* **Response2** is better overall, especially in terms of Medical accuracy Scenario relevance and Practical emergency and recovery guidance

## Data Preparation for RAG

In [31]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

### Loading the Data

In [33]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [34]:
merck_pdf_path = '/content/drive/MyDrive/medical_diagnosis_manual.pdf'
pdf_loader = PyMuPDFLoader(merck_pdf_path)
merck_pdf = pdf_loader.load()

### Data Overview

#### Checking the first 5 pages

In [35]:
for i in range(5):
    print(f"Page Number : {i+1}",end="\n")
    print(merck_pdf[i].page_content,end="\n")

Page Number : 1
desai.sana1@gmail.com
B0PAGYZ4X1
ant for personal use by desai.sana1@gm
shing the contents in part or full is liable 

Page Number : 2
desai.sana1@gmail.com
B0PAGYZ4X1
This file is meant for personal use by desai.sana1@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.

Page Number : 3
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    ..........................................................................................................................................................................................

#### Checking the number of pages

In [36]:
print("length of merck PDF is-",len(merck_pdf))

length of merck PDF is- 4114


### Data Chunking

In [37]:
# Initializing a RecursiveCharacterTextSplitter to split the text into manageable chunks for embedding and retrieval
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 20
)

In [38]:
#loading the PDF document, extracting its text, and splitting it into smaller chunks
document_chunks = pdf_loader.load_and_split(text_splitter)

In [39]:
#Checking the number of text chunks the pdf has been split into
print("number of text chunks the pdf has been split into-",len(document_chunks))

number of text chunks the pdf has been split into- 8487


Confirming that there is overlap between chunks

In [40]:
document_chunks[0].page_content

'desai.sana1@gmail.com\nB0PAGYZ4X1\nant for personal use by desai.sana1@gm\nshing the contents in part or full is liable'

In [41]:
document_chunks[1].page_content

'desai.sana1@gmail.com\nB0PAGYZ4X1\nThis file is meant for personal use by desai.sana1@gmail.com only.\nSharing or publishing the contents in part or full is liable for legal action.'

In [42]:
document_chunks[2].page_content

'Table of Contents\n1\nFront    ................................................................................................................................................................................................................\n1\nCover    .......................................................................................................................................................................................................\n2\nFront Matter    ...........................................................................................................................................................................................\n53\n1 - Nutritional Disorders    ...............................................................................................................................................................\n53\nChapter 1. Nutrition: General Considerations    ...........................................................................................

**Observation-**

As expected, there are some overlaps:
* The last 20 tokens (not characters) of document_chunks[0]
appear at the beginning of document_chunks[1].
* Last sentence fragment of chunk 0 occurs in the First word of chunk 1:
* If we increase the chunk_overlap, the overlapping length of the sentence will also increase.

### Embedding

In [43]:
# This model is chosen because it supports a maximum input length of 512 tokens,
# which matches our chunk size during document splitting. It also produces high-quality
# 1024-dimensional embeddings suitable for semantic search and RAG pipelines.
embedding_model = SentenceTransformerEmbeddings(model_name='thenlper/gte-large')

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

In [44]:
# Generating embedding for the first document chunk
embedding_1 = embedding_model.embed_query(document_chunks[0].page_content)
# Generating embedding for the second document chunk
embedding_2 = embedding_model.embed_query(document_chunks[1].page_content)

In [45]:
#Checking if both are of the same size
print("Dimension of the embedding vector ",len(embedding_1))
len(embedding_1)==len(embedding_2)

Dimension of the embedding vector  1024


True

In [46]:
embedding_1,embedding_2

([-0.02141253463923931,
  -0.004343502223491669,
  -0.002323613502085209,
  -0.011759995482861996,
  -0.014929787255823612,
  -0.026302766054868698,
  -0.013058159500360489,
  0.030034080147743225,
  0.008949949406087399,
  0.01830601878464222,
  0.02505834400653839,
  0.004774370230734348,
  0.017779041081666946,
  -0.03663885220885277,
  -0.00422451039776206,
  -0.01146604958921671,
  0.0014619737630710006,
  -0.019188594073057175,
  -0.00695505365729332,
  0.004877684637904167,
  0.010211425833404064,
  0.010608792304992676,
  -0.07724888622760773,
  -0.02444807067513466,
  -0.009776102378964424,
  0.03191791847348213,
  0.0024320296943187714,
  -0.007021660450845957,
  0.07554978132247925,
  0.046516094356775284,
  -0.01974877528846264,
  -0.0022843307815492153,
  0.039009422063827515,
  -0.0476209819316864,
  -0.02252781391143799,
  -0.0010169813176617026,
  0.057295240461826324,
  -0.03271351009607315,
  -0.02905232273042202,
  -0.04068784415721893,
  0.023215264081954956,
  0.00


* The embedding model provides a fixed-length vector for any number of chunks.
* This is necessary because we want to compare them for similarity.


### Vector Database

In [47]:
import os
import time


In [48]:
# Creating the output directory 'merck_db' if it doesn't already exist, so we can save the processed data or vector database files there.
out_dir = 'merck_db'

if not os.path.exists(out_dir):
  os.makedirs(out_dir)

In [50]:
start = time.time()
document_chunks = document_chunks[:100]

In [51]:
vectorstore = Chroma.from_documents( #creating a Chroma vector store from a set of document chunks.
    document_chunks, #creating a list of text chunks that will be converted into embeddings..
    embedding_model, #model responsible for embedding the document chunks into vector representations
    persist_directory=out_dir #name of the collection in the Chroma database
)
print("Execution time:", time.time() - start)

Execution time: 11.103809356689453


In [52]:
#Loading Chroma vector store with the given embedding model
vectorstore = Chroma(persist_directory=out_dir, embedding_function=embedding_model)

In [53]:
#Accessing the embedding function used in the Chroma vector store
vectorstore.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False})
  (2): Normalize()
), model_name='thenlper/gte-large', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [54]:
#Performing a similarity search in the vector store to find the top 3 most similar documents to "Alopecia Areata"
vectorstore.similarity_search("Alopecia Areata ",k=3)

[Document(page_content='742\n7 - Dermatologic Disorders    ....................................................................................................................................................\n742\nChapter 71. Approach to the Dermatologic Patient    .......................................................................................................\n755\nChapter 72. Principles of Topical Dermatologic Therapy    ............................................................................................\n760\nChapter 73. Acne & Related Disorders    ...............................................................................................................................\n766\nChapter 74. Bullous Diseases    .................................................................................................................................................\n771\nChapter 75. Cornification Disorders    ......................................................................

In [55]:
#Performing a similarity search in the vector store to find the top  most similar documents to "Alopecia Areata"
vectorstore.similarity_search("Alopecia Areata ",k=1)

[Document(page_content='742\n7 - Dermatologic Disorders    ....................................................................................................................................................\n742\nChapter 71. Approach to the Dermatologic Patient    .......................................................................................................\n755\nChapter 72. Principles of Topical Dermatologic Therapy    ............................................................................................\n760\nChapter 73. Acne & Related Disorders    ...............................................................................................................................\n766\nChapter 74. Bullous Diseases    .................................................................................................................................................\n771\nChapter 75. Cornification Disorders    ......................................................................

### Retriever

In [56]:
retriever = vectorstore.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 2} #Retrieving the top 2 most similar documents for a given query.
)

In [57]:
user_input = 'What are the symptoms of migraine?'
rel_docs = retriever.get_relevant_documents(user_input)
rel_docs

/usr/local/lib/python3.11/dist-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 0.3.0. Use invoke instead.
  warn_deprecated(


[Document(page_content="Testing for Hepatic and Biliary Disorders;\nAlcoholic Liver Disease; Fibrosis and Cirrhosis;\nVascular Disorders of the Liver; Gallbladder\nand Bile Duct Disorders\nSTEWART SHANKEL, MD\nClinical Professor of Medicine and Director\nof Clinical Instruction, University of California,\nRiverside\nUrinary Tract Infections\nWILLIAM R. SHAPIRO, MD\nProfessor of Clinical Neurology, University\nof Arizona College of Medicine, Tucson;\nChief, Neuro-oncology, Barrow Neurological\nInstitute\nIntracranial and Spinal Tumors\nMICHAEL J. SHEA, MD\nProfessor of Internal Medicine, Section Head,\nCardiovascular Medicine Outpatient Services\nand Clinical Evaluation, University of\nMichigan Health Systems\nCardiovascular Tests and Procedures\nDAVID D. SHERRY, MD\nProfessor of Pediatrics, University of\nPennsylvania; Director, Clinical\nRheumatology, The Children's Hospital of\nPhiladelphia\nRheumatic Fever; Bone and Connective Tissue\nDisorders in Children\nSTEPHEN D. SILBERSTEIN, M

In [58]:
model_output = llm (
    prompt="what is Bromhidrosis",
    max_tokens=4500,
    temperature=10
)
print(model_output)

Llama.generate: prefix-match hit


{'id': 'cmpl-e4c4a382-b5f7-4ff3-b77c-ec433c145005', 'object': 'text_completion', 'created': 1754758455, 'model': '/root/.cache/huggingface/hub/models--TheBloke--Mistral-7B-Instruct-v0.2-GGUF/snapshots/3a6fbf4a41a1d52e415a4958cde6856d34b2db93/mistral-7b-instruct-v0.2.Q6_K.gguf', 'choices': [{'text': ', How do we develop this smell condition?\nBromhidrosis is a medical condition that causes a particularly foul body odor. This odor is the result of bacterial action on sweat that hasn\'t been washed away properly.\nThe bacteria in question are usually found in the armpits and the genital area, as these are areas where sweat collects and where it\'s more difficult to keep clean through daily washing. The bacteria feed on the sweat, which contains amino acids and lipids, producing unpleasant smelling byproducts.\nBromhidrosis can be divided into two types based on when it occurs:\n1. Odistichiasis: This type is also known as "axillary bromhidrosis," and it affects the armpits. The odor usual

### System and User Prompt Template

In [59]:
# System message instructing the LLM to only answer using Merck Manual 19th Edition
qna_system_message = """
You are a helpful assistant trained to answer questions based only on the Merck Manual of Medical Diagnosis and Therapy, Nineteenth Edition.
Use the context provided to find accurate and reliable answers.
If the answer is not found in the context, reply with "I don't know".
Do not mention the context or the Merck Manual in your final answer.

You are an AI assistant specializing in medical knowledge. Your role is to provide clear, precise, and medically reliable responses based on established medical guidelines and best practices.

When answering, prioritize factual correctness, align with widely accepted medical standards, and ensure clarity for both medical professionals and general users.
If a query requires specific reference materials beyond general medical knowledge, acknowledge the limitation rather than speculating.
"""

In [60]:
# Template for formatting the user's input with context from the Merck Manual, 19th Edition and the actual medical question.
qna_user_message_template = """
###Context
The following excerpts are from the Merck Manual of Medical Diagnosis and Therapy, Nineteenth Edition:
{context}

###Question
{question}
"""

### Response Function

In [61]:
def generate_rag_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [62]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response(user_input1))

Llama.generate: prefix-match hit


Based on the context provided, the Merck Manual includes information on sepsis and septic shock under Chapter 244.7, Sepsis & Septic Shock (p. 2448). The protocol for managing sepsis in a critical care unit involves early recognition, prompt initiation of appropriate antibiotic therapy, fluid resuscitation, and supportive measures such as mechanical ventilation and vasopressor support if necessary.

1. Recognition: Identify patients at risk based on clinical signs and laboratory findings (fever, tachycardia, tach


Fine Tuning-max_tokens=370

In [64]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response(user_input1,max_tokens=370))

Llama.generate: prefix-match hit


Based on the context provided, the Merck Manual includes information on sepsis and septic shock under Chapter 244.7, Sepsis & Septic Shock (p. 2448). The protocol for managing sepsis in a critical care unit involves early recognition, prompt initiation of appropriate antibiotic therapy, fluid resuscitation, and supportive measures such as mechanical ventilation and vasopressor support if necessary.

1. Recognition: Identify patients at risk based on clinical signs and laboratory findings (fever, tachycardia, respiratory distress, altered mental status, lactic acidosis, leukocytosis or leukopenia, thrombocytopenia, and microbiological evidence of infection).
2. Antibiotic therapy: Administer broad-spectrum antibiotics based on the suspected source of infection and local microbial resistance patterns.
3. Fluid resuscitation: Aim for a mean arterial pressure (MAP) ≥65 mm Hg and a central venous pressure (CVP) 8-12 cm H2O in patients with septic shock. Use crystalloids, colloids, or blood 

**Observations:**

* **Response1** is too high-level and incomplete.

* **Response2** is more clinically actionable, structured, and aligns better with actual sepsis protocols.

  The truncation in both is likely due to the max_tokens=370 limit — increasing this would allow a full protocol output.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [65]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
print(generate_rag_response(user_input2))

Llama.generate: prefix-match hit


Based on the context provided, the Merck Manual does not contain specific information about the treatment of appendicitis with medication. However, it is mentioned that appendicitis typically presents with symptoms such as periumbilical pain, anorexia, nausea, and vomiting, which may be followed by rebound tenderness and guarding in the right lower quadrant (RLQ) on physical examination. The manual also states that appendicitis is usually treated surgically, specifically with an appendectomy. Therefore, the answer would be:

Appendicitis is characterized by symptoms such as per


Fine Tuning - Retreiver Parameters
k = 2

In [66]:
retriever2 = vectorstore.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 2} #Retrieving the top 2 most similar documents for a given query.
)

In [67]:
def generate_rag_response2(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever2.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

In [68]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
print(generate_rag_response2(user_input2))

Llama.generate: prefix-match hit


###Answer
The common symptoms for appendicitis include sudden onset of pain in the lower right abdomen, loss of appetite, nausea and vomiting, constipation or diarrhea, fever, and a feeling of being sick. Appendicitis cannot be cured via medicine alone as it requires surgical removal of the appendix to prevent rupture and potential complications such as peritonitis. The standard surgical procedure for treating appendicitis is an appendectomy, which can be performed laparoscopically or open depending on the severity of the case.


**Observation:**
* **Response1** is more source-driven but incomplete and truncated.

* **Response2** is clearer, more complete, and more actionable for both medical and lay readers.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

Basic parameters

In [69]:
user_input3 = 'What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'
print(generate_rag_response(user_input3))

Llama.generate: prefix-match hit


Based on the provided context, the relevant chapters to consider for answering this question are Chapters 76 (Dermatitis) and 86 (Hair Disorders).

First, let's discuss potential causes:
1. Dermatitis: This is a common skin condition that can cause patchy hair loss. Various types of dermatitis include contact dermatitis, seborrheic dermatitis, and alopecia areata. Contact dermatitis results from an allergic or irritant reaction to substances like cosmetics, metals


Fine Tuning - LLM Parameters
Temperature = 0.7

In [70]:
user_input3 = 'What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'
print(generate_rag_response(user_input3, temperature=0.7))

Llama.generate: prefix-match hit


Based on the context provided, Chapter 76 of the Merck Manual covers Dermatitis. While there isn't a specific mention of "sudden patchy hair loss" in the given text, dermatitis can cause hair loss as a symptom. The following are some causes and treatments for common types of dermatitis:

1. **Contact Dermatitis**: This type is caused by an allergic or irritant reaction to certain substances that come into contact with the skin. Common causes include cosmetics, hair dyes, metals (like nickel),


**Observation**
* **Response1** is broader (two chapters, more conditions) but less explicit about context limitations.

* **Response2** is more cautious and structured but narrower in scope.

  A combined approach — acknowledging the context gap like Response 2, but including the wider condition coverage from Response 1 — would give the best result.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

Basic parameters

In [71]:
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
print(generate_rag_response(user_input4))

Llama.generate: prefix-match hit


Based on the context provided, the relevant chapters are 324 (Traumatic Brain Injury), 174 (Coma & Impaired Consciousness), and 175 (Delirium & Dementia). The Merck Manual discusses various aspects of brain injuries, including diagnosis, prognosis, and treatment.

The treatments for a person with a brain injury may include:

1. Supportive care: This includes maintaining adequate ventilation, controlling intracranial pressure, managing fluids and electrolytes, providing nutrition, and preventing complications such


Fine Tuning - LLM Parameters
Top_p = 0.8

In [72]:
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
print(generate_rag_response(user_input4, top_p=0.8))

Llama.generate: prefix-match hit


Based on the context provided, the relevant chapters are 324 (Traumatic Brain Injury), 174 (Coma & Impaired Consciousness), and 175 (Delirium & Dementia). The Merck Manual discusses various aspects of brain injuries, including diagnosis, prognosis, and treatment.

The treatments for a person with a brain injury may include:

1. Supportive care: This includes maintaining adequate ventilation, controlling intracranial pressure, managing fluid and electrolyte balance, providing nutritional support, and preventing


**Observation**
* Both are nearly identical in structure and content.

* Both are cut off prematurely — likely due to max_tokens limit or early stop in generation.

* Neither fully covers the complete treatment spectrum for brain injury.

* Top_p = 0.8 setting in Response 2 doesn’t seem to have affected diversity since both are near duplicates.



### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

Basic Parameters

In [73]:
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
print(generate_rag_response(user_input5))

Llama.generate: prefix-match hit


Based on the context provided, I do not find specific information regarding the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. However, I can provide some general information about leg fractures and their management from Chapter 349 (Care of the Surgical Patient) in the Merck Manual:

1. Immobilize the affected limb using a splint or cast to prevent further injury and promote healing.
2. Apply ice packs intermittently to reduce swelling and pain.
3. Administer pain medication as prescribed


### Fine-tuning

Fine Tuning - LLM Parameters
Top_k = 25

In [74]:
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
print(generate_rag_response(user_input5, top_k=25))

Llama.generate: prefix-match hit


Based on the context provided, I do not find specific information regarding the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. However, I can provide some general information about leg fractures and their care from Chapters 349 (Care of the Surgical Patient) and 350 (Rehabilitation).

1. Precautions:
   - Maintain immobilization to prevent further injury or displacement of the fracture.
   - Monitor for signs of complications such as infection, nerve damage, or compartment


**Observation**
* **Response1** Very general, no explanation of when to seek advanced care or specifics on splinting.
* **Response2** Adds more medically relevant advice (complication monitoring).
Still lacks treatment details such as elevation, transport safety, and fracture stabilization technique.
* **Response1** is short and overly general.

* **Response2** is more comprehensive but still incomplete due to truncation.

## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

Rating System

In [75]:
groundedness_rater_system_message  = """

You will be presented a ###Question, ###Context used by the AI system and AI generated ###Answer.

Your task is to judge the extent to which the ###Answer is derived from ###Context.

Rate it 1 - if The ###Answer is not derived from the ###Context at all
Rate it 2 - if The ###Answer is derived from the ###Context only to a limited extent
Rate it 3 - if The ###Answer is derived from ###Context to a good extent
Rate it 4 - if The ###Answer is derived from ###Context mostly
Rate it 5 - if The ###Answer is is derived from ###Context completely

Please note: Make sure you give a single overall rating in the range of 1 to 5 along with an overall explanation.

Instructions:
1. First write down the steps that are needed to evaluate the answer as per the metric.
2. Give a step-by-step explanation if the answer adheres to the metric considering the question and context as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the answer using the evaluation criteria and assign a score.

"""

In [76]:
relevance_rater_system_message = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
Relevance measures how well the answer addresses the main aspects of the question, based on the context.
Consider whether all and only the important aspects are contained in the answer when evaluating relevance.

Instructions:
1. First write down the steps that are needed to evaluate the context as per the metric.
2. Give a step-by-step explanation if the context adheres to the metric considering the question as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the context using the evaluation criteria and assign a score.
"""

In [77]:
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [78]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            echo=False
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            echo=False
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            echo=False
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [87]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
ground,rel = generate_ground_relevance_response(user_input1,max_tokens=370)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 To evaluate the answer, we need to follow these steps:

1. Identify the key information in the context related to sepsis management in a critical care unit.
2. Compare this information with the steps mentioned in the answer.
3. Determine the extent to which the answer is derived from the context.

Step 1: The context provides several chapters related to various aspects of critical care, including one on shock and fluid resuscitation (Chapter 226), which discusses sepsis and septic shock in detail.

Step 2: The answer mentions the following steps for managing sepsis in a critical care unit:

   a. Recognize signs and symptoms of sepsis
   b. Obtain diagnostic tests
   c. Initiate fluid resuscitation
   d. Administer antibiotics
   e. Provide adequate oxygenation and ventilation
   f. Monitor and manage hemodynamic instability
   g. Correct electrolyte imbalances, acid-base disturbances, and coagulation abnormalities
   h. Provide source control for the infection
   i. Monitor and manag

**Observation:**

* Groundedness (Score: 4/5)

  Stepwise structure: Clearly outlines the evaluation steps (identify, compare, determine extent).
  Detailed mapping: Lists each step in the answer (a–i) and matches them to context content (e.g., Chapter 226).
  Specific reference: Mentions an actual chapter number where relevant content is located.

  Logical conclusion: The 4/5 rating is justified with coverage alignment.

* Relevance (Context Score: 3/5)

  Defines what to look for (main aspects, completeness, detail level).
  Notes that the context is rich in related information but lacks an explicit protocol.
  
  Honest limitation: Acknowledges that the context sets up the answer but doesn’t fully answer the question directly.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [83]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
ground,rel = generate_ground_relevance_response(user_input2,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 To evaluate the answer, we need to follow these steps:

1. Identify the key information in the question and context.
2. Determine if the answer addresses the question directly.
3. Evaluate how well the answer uses information from the context.

The question asks about the common symptoms of appendicitis and whether it can be cured via medicine or if surgery is required. The context provided is a list of chapters from a medical textbook, with no direct relation to the question.

The answer provides the common symptoms for appendicitis and states that it cannot be cured via medicine alone and requires surgery (appendectomy) for treatment.

Based on this information, we can evaluate the answer as follows:

1. The answer directly addresses the question by providing the common symptoms of appendicitis and explaining that it cannot be cured with medicine alone and requires surgery.
2. The answer uses information from the context indirectly, as there is no specific context related to appendi

**Observations:**
* Groundedness (Score: 3/5)

  Structured approach: Clearly lists steps for evaluation before scoring.
  Direct question addressing: Confirms that the answer gives both the symptoms of appendicitis and the treatment approach (surgery).

* Relevance (Context Score: 2/5)

  Clear reasoning: Notes explicitly that the context contains chapters on digestive system disorders, which may include appendicitis.

  Honest limitation acknowledgment: States the context does not directly cover symptoms or treatment specifics.



### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [84]:
user_input3 = 'What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'
ground,rel = generate_ground_relevance_response(user_input3,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 To evaluate the answer, we need to follow these steps:

1. Identify the key information in the context related to the question.
2. Determine if the answer mentions and correctly interprets the key information from the context.
3. Evaluate the extent to which the answer goes beyond the context, if any.

Firstly, the key information in the context is that Chapter 86 of The Merck Manual discusses hair disorders.

Secondly, the answer correctly identifies this information and adds that sudden patchy hair loss can be caused by various conditions such as alopecia areata, telogen effluvium, or traction alopecia. It also mentions some treatments for these conditions based on The Merck Manual.

Thirdly, the answer goes beyond the context to a good extent by providing more detailed information about each condition and their respective treatments. However, it does not provide any new or incorrect information.

Therefore, I would rate the answer as a 4 because it is derived from the context mostl

**Observations:**

* Groundedness ( Score: 4/5)

  Structured process: Lists explicit evaluation steps before giving the score.

  Clear reasoning: Notes that the answer is mostly from the context but also includes relevant additions.

  No hallucinations detected: States that added details are correct and relevant.

* Relevance( Score: Mostly Relevant)

  Topic alignment: Clearly matches question scope (“causes and treatments for sudden patchy hair loss”).

  Detailed breakdown: Lists both causes and treatments as found in the context.

  No irrelevant info: Explicitly states no unrelated content is present.

### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [85]:
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
ground,rel = generate_ground_relevance_response(user_input4,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 To evaluate the answer, we need to follow these steps:

1. Identify the key concepts from the question and context. In this case, the key concepts are "person," "brain injury," "temporary or permanent impairment of brain function," and "treatments."
2. Determine if the answer addresses all the key concepts from the question. The answer does address the person with a brain injury and the temporary or permanent impairment of brain function, as well as the recommended treatments.
3. Check if the information in the answer is derived directly from the context. The answer mentions "Chapter 324" for the treatment recommendations, which is indeed present in the context. It also refers to "Chapters 174, 180, and 182" for additional information on brain injuries and their treatments, which are also present in the context.

Based on this evaluation, I would rate the answer as a 5 because it is completely derived from the context. The answer accurately references specific chapters in the context 

**Observations:**
* Groundedness (rating = 5/5)

  Cites specific chapters and claims the answer is drawn from them.
* Relevance (rating = 4/5)

  Identifies the main ask (treatments for TBI) and says the context covers it.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [86]:
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
ground,rel = generate_ground_relevance_response(user_input5,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 The answer provided is derived from the context to a good extent (rating: 3). The context includes chapters related to "Care of the Surgical Patient" and "Rehabilitation," which directly address the necessary precautions and treatment steps for a person with a fractured leg. The answer covers several aspects, such as immobilization, pain management, positioning, nutrition, infection prevention, mobility assistance, and rehabilitation services, all of which are discussed in the context. However, some parts of the answer, like psychological support and return-to-work considerations, are not explicitly mentioned in the context but are generally applicable to patients with injuries or health conditions.

 To evaluate the context as per the relevance metric, the following steps are needed:

1. Identify the main aspects of the question: necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and re

**Observations:**
* Groundedness: 2–3/5 (depending on whether those rehab/psych sections truly exist in context).

* Relevance: 3–4/5 (good topical alignment, but weak evidence + truncation).

## Actionable Insights and Business Recommendations

**Business Insights**

* **Accuracy and Trustworthiness**

  RAG-based answers deliver context-specific, reliable, and trustworthy information, enhancing customer confidence and satisfaction in medical applications.

* **Enhanced User Experience**

  The RAG model provides personalized, contextually relevant, and highly accurate responses, improving overall user engagement.

* **Cost Optimization in Medical Consultations**

  By automating medical Q&A, the model reduces the need for routine consultations, lowering costs in telemedicine and healthcare support services.

**Recommendations**

* **Ongoing Fine-Tuning**

  Regularly update and fine-tune the model to align with the latest medical guidelines and research findings.

* **Ethical Compliance**

  Prioritize data privacy and security to ensure adherence to regulations such as HIPAA and GDPR.

* **Specialized Model Development**

  Create dedicated models for specific medical domains such as orthopedics, gastroenterology, and oncology to further enhance accuracy and relevance.

<font size=6 color='blue'>Power Ahead</font>
___